Add velocity dispersion function sampler to skypy

Vikramb1 opened the pull request:
## Description
Add a velocity dispersion function sampler to skypy, which is describes the populations of lenses in sigma_v space. Sigma_v is velocity dispersion of the lens. The equation used for this is found in [Choi et al. (2017)](https://iopscience.iop.org/article/10.1086/511060/pdf). The paper uses data of Elliptical galaxies from the Sloan Digital Sky survey. Closes #575.
## Checklist
- [x] Follow the [Contributor Guidelines](https://github.com/skypyproject/skypy/blob/main/CONTRIBUTING.rst)
- [x] Write unit tests
- [x] Write documentation strings
- [ ] Assign someone from your working team to review this pull request
- [ ] Assign someone from the infrastructure team to review this pull request

sibirrer commented:
@Lucia-Fonseca @itrharrison  or others, is this PR in review or let us know if and how we can proceed. Thank you!

philipp128 commented:
Thanks @Vikramb1 for the PR. Could you add a description for this PR and link to an open issue which this PR is trying to resolve? If no issue exist, could you open one? This would help to understand what this PR is adding and makes the review process easier.

Vikramb1 commented:
> Thanks @Vikramb1 for the PR. Could you add a description for this PR and link to an open issue which this PR is trying to resolve? If no issue exist, could you open one? This would help to understand what this PR is adding and makes the review process easier.

Hi Philipp, just done that now. I am a bit new to all this, so let me know if there is anything you want to change or for me to explain better.

philipp128 commented:
> > Thanks @Vikramb1 for the PR. Could you add a description for this PR and link to an open issue which this PR is trying to resolve? If no issue exist, could you open one? This would help to understand what this PR is adding and makes the review process easier.
> 
> Hi Philipp, just done that now. I am a bit new to all this, so let me know if there is anything you want to change or explain better.

Thanks, much better! One more thing though, it is always helpful to link to the corresponding issue. We normally do this by adding "Closes #Issue_number". This has the advantage of automatically closing the issue if the PR is merged.

I also approved the contribution such that the tests can run. Note that the flake8 test failed, which is checking for code style. As a first step, can you try to fix this? You can install flake8 on your machine and just run it with `flake8 path/to/file`.

Vikramb1 commented:
> > > Thanks @Vikramb1 for the PR. Could you add a description for this PR and link to an open issue which this PR is trying to resolve? If no issue exist, could you open one? This would help to understand what this PR is adding and makes the review process easier.
> > 
> > 
> > Hi Philipp, just done that now. I am a bit new to all this, so let me know if there is anything you want to change or explain better.
> 
> Thanks, much better! One more thing though, it is always helpful to link to the corresponding issue. We normally do this by adding "Closes #Issue_number". This has the advantage of automatically closing the issue if the PR is merged.
> 
> I also approved the contribution such that the tests can run. Note that the flake8 test failed, which is checking for code style. As a first step, can you try to fix this? You can install flake8 on your machine and just run it with `flake8 path/to/file`. Please check both files, i.e. also the test file. Thanks!

Ah, ok. I added a reference to the issue and I think the code should pass the flake8 test now. Thanks!

Vikramb1 commented:
I noticed the coverage test fail, and I think the latest commit should solve that.

Vikramb1 commented:
Is there anything else you would like me to do for this pr @philipp128 ?

philipp128 commented on skypy/galaxies/vdf_sampler.py:
```diff
@@ -0,0 +1,92 @@
+
+
+def sample_vdf(x_min, x_max, resolution=100, size=1):
+    """Sample from velocity dispersion function of elliptical galaxies in the local universe [1]_.
```
Please add the equation in LaTeX style to the docstring in a `Note` section and refer to the equation number of the paper. This makes it easier to understand which equation is implemented. See for example docstring of [skypy.galaxies.redshift.py.smail](https://github.com/skypyproject/skypy/blob/df2536974c122dd72498f8d146913c14c832833b/skypy/galaxies/redshift.py#L27).

Vikramb1 replied:
Done.

philipp128 commented on skypy/galaxies/vdf_sampler.py:
```diff
@@ -0,0 +1,92 @@
+from scipy.interpolate import interp1d
+
+
+def sample_vdf(x_min, x_max, resolution=100, size=1):
```
I suggest not using `sample` in the name of the function. We have similar functions for luminosities and redshifts that sample from a Schechter function. See main comment.

Vikramb1 replied:
Done.

philipp128 commented on skypy/galaxies/vdf_sampler.py:
```diff
@@ -0,0 +1,92 @@
+from scipy.interpolate import interp1d
+
+
+def sample_vdf(x_min, x_max, resolution=100, size=1):
```
Please do not use the variable `x` here. I suggest `vd_min`/`vd_max` or just `sigma_min` and `sigma_max`. I prefer `vd` as it makes it a bit clearer. 

Also similar to the examples the default of size should be `None` and in which case the function should return one value.

Vikramb1 replied:
Done.

philipp128 commented on skypy/galaxies/vdf_sampler.py:
```diff
@@ -0,0 +1,92 @@
+def sample_vdf(x_min, x_max, resolution=100, size=1):
+    """Sample from velocity dispersion function of elliptical galaxies in the local universe [1]_.
+
+    Parameters
```
Update according to the other comments.

Vikramb1 replied:
> From the context I assume that you implement a sampler based on equation (4) of the Choi et al. paper. Please clarify this in the docstring.
> 
> I see that the velocity dispersion is simply a Schechter function, which is already implemented in SkyPy. I suggest that you follow the same syntax. I.e. add a module `velocity_dispersion.py` to the galaxy package and in this module add a function `schechter_vdf()`. You can find examples in [skypy.galaxies.luminosity](https://github.com/skypyproject/skypy/blob/main/skypy/galaxies/luminosity.py), [skypy.galaxies.redshift](https://github.com/skypyproject/skypy/blob/main/skypy/galaxies/redshift.py) and [skypy.galaxies.stellar_mass](https://github.com/skypyproject/skypy/blob/main/skypy/galaxies/stellar_mass.py).
> 
> We have also already implemented a sampler `schechter` to sample from the Schechter function in [skypy.utils.random](https://github.com/skypyproject/skypy/blob/df2536974c122dd72498f8d146913c14c832833b/skypy/utils/random.py#L21). You can use it in the same way as done in the examples mentioned above. In this case you don't need the class `PDFSampling` and the function `approx_cdf_1d()`.
> 
> I added some further comments about naming conventions at the specific lines in the codes.

Hi Philipp - thanks for the feedback. I have committed the changes. Please let me know what you think.

philipp128 commented on skypy/galaxies/vdf_sampler.py:
```diff
@@ -0,0 +1,92 @@
+
+    Returns
+    -------
+    x_sample: array_like
```
```suggestion
    velocity_dispersion: array_like
```

philipp128 commented on skypy/galaxies/vdf_sampler.py:
```diff
@@ -0,0 +1,92 @@
+    Returns
+    -------
+    x_sample: array_like
+        Samples drawn from vdf function.
```
Sth like: "Velocity dispersion drawn from Schechter function"

Vikramb1 replied:
Done.

philipp128 commented on skypy/galaxies/vdf_sampler.py:
```diff
@@ -0,0 +1,92 @@
+    ----------
+    .. [1] Choi, Park and Vogeley, (2007), astro-ph/0611607, doi:10.1086/511060
+    """
+    x = np.linspace(0, 1000, resolution)
+
+    def vdf_func(x):
+        return 8e-3*(x/161)**2.32*np.exp(-(x/161)**2.67)*(2.67/special.gamma(2.32/2.67))*(1/x)
+
+    y = vdf_func(x)
+    pdf_sampler = PDFSampling(x, y[1:], x_min, x_max)
+    test_draw = pdf_sampler.draw_one()
+    assert len(test_draw) == 1
+    return pdf_sampler.draw(size)
```
See main comment.

philipp128 commented on skypy/galaxies/vdf_sampler.py:
```diff
@@ -0,0 +1,92 @@
+        """
+        :param bin_edges: bin edges of PDF values
+        :param pdf_array: pdf array of given bins (len(bin_edges)-1)
+        """
+        assert len(bin_edges) == len(pdf_array) + 1
+        self.x_min, self.x_max = x_min, x_max
+        self._cdf_array, self._cdf_func, self._cdf_inv_func = approx_cdf_1d(bin_edges, pdf_array)
+
+    def draw(self, n=1):
+        """
+        :return:
+        """
+        t_min, t_max = self._cdf_func(self.x_min), self._cdf_func(self.x_max)
+        p = np.random.uniform(t_min, t_max, n)
+        return self._cdf_inv_func(p)
+
+    @property
+    def draw_one(self):
+        """
+        :return:
+        """
+        return self.draw(n=1)
+
+
+def approx_cdf_1d(bin_edges, pdf_array):
+
+    """
+    :param bin_edges: bin edges of PDF values
+    :param pdf_array: pdf array of given bins (len(bin_edges)-1)
+    :return: cdf, interp1d function of cdf, inverse interpolation function
+    """
+    assert len(bin_edges) == len(pdf_array) + 1
+    norm_pdf = pdf_array/np.sum(pdf_array)
+    cdf_array = np.zeros_like(bin_edges)
+    cdf_array[0] = 0
+    for i in range(0, len(norm_pdf)):
+        cdf_array[i+1] = cdf_array[i] + norm_pdf[i]
+    cdf_func = interp1d(bin_edges, cdf_array)
+    cdf_inv_func = interp1d(cdf_array, bin_edges)
+    return cdf_array, cdf_func, cdf_inv_func
```
See main comment.

philipp128 commented on skypy/galaxies/tests/test_vdf_sampler.py:
```diff
@@ -0,0 +1,7 @@
+import numpy as np
+
+
+def test_vdf_sampler():
```
Add some more tests. Also use the test functions of the examples in the main comment as an example.

Vikramb1 replied:
Done.

philipp128 reviewed (changes requested):
From the context I assume that you implement a sampler based on equation (4) of the Choi et al. paper. Please clarify this in the docstring.

I see that the velocity dispersion is simply a Schechter function, which is already implemented in SkyPy. 
I suggest that you follow the same syntax. I.e. add a module `velocity_dispersion.py` to the galaxy package and in this module add a function `schechter_vdf()`. You can find examples in [skypy.galaxies.luminosity](https://github.com/skypyproject/skypy/blob/main/skypy/galaxies/luminosity.py), [skypy.galaxies.redshift](https://github.com/skypyproject/skypy/blob/main/skypy/galaxies/redshift.py) and [skypy.galaxies.stellar_mass](https://github.com/skypyproject/skypy/blob/main/skypy/galaxies/stellar_mass.py).

We have also already implemented a sampler `schechter` to sample from the Schechter function in [skypy.utils.random](https://github.com/skypyproject/skypy/blob/df2536974c122dd72498f8d146913c14c832833b/skypy/utils/random.py#L21). You can use it in the same way as done in the examples mentioned above. In this case you don't need the class `PDFSampling` and the function `approx_cdf_1d()`.

I added some further comments about naming conventions at the specific lines in the codes.

Vikramb1 commented:
> From the context I assume that you implement a sampler based on equation (4) of the Choi et al. paper. Please clarify this in the docstring.
> 
> I see that the velocity dispersion is simply a Schechter function, which is already implemented in SkyPy. I suggest that you follow the same syntax. I.e. add a module `velocity_dispersion.py` to the galaxy package and in this module add a function `schechter_vdf()`. You can find examples in [skypy.galaxies.luminosity](https://github.com/skypyproject/skypy/blob/main/skypy/galaxies/luminosity.py), [skypy.galaxies.redshift](https://github.com/skypyproject/skypy/blob/main/skypy/galaxies/redshift.py) and [skypy.galaxies.stellar_mass](https://github.com/skypyproject/skypy/blob/main/skypy/galaxies/stellar_mass.py).
> 
> We have also already implemented a sampler `schechter` to sample from the Schechter function in [skypy.utils.random](https://github.com/skypyproject/skypy/blob/df2536974c122dd72498f8d146913c14c832833b/skypy/utils/random.py#L21). You can use it in the same way as done in the examples mentioned above. In this case you don't need the class `PDFSampling` and the function `approx_cdf_1d()`.
> 
> I added some further comments about naming conventions at the specific lines in the codes.

Hi Philipp - thanks for the feedback. I have committed the changes. Please let me know what you think.

philipp128 commented on skypy/galaxies/velocity_dispersion.py:
```diff
@@ -0,0 +1,69 @@
+
+import numpy as np
+import scipy.special as special
+
```
```suggestion

__all__ = [
    'schechter_vdf',
]

```

philipp128 replied:
Important for the docs.

philipp128 commented on skypy/galaxies/velocity_dispersion.py:
```diff
@@ -0,0 +1,69 @@
+    The probability distribution function :math:`p(\sigma)` for redshift :math:`\sigma`
+    is given by Choi et al. [1]_ as
+
+    .. math::
+        \begin{eqnarray}&&{dn}={\phi }_{*}{(\displaystyle \frac{\sigma }{{\sigma }_{*}})}^{\alpha }
+        \mathrm{exp}[-{(\displaystyle \frac{\sigma }{{\sigma }_{*}})}^{\beta }]\displaystyle
+        \frac{\beta }{{\rm{\Gamma }}(\alpha /\beta )}\displaystyle \frac{1}{\sigma },
+        \end{eqnarray}\tag{3} \;.
```
Can you make this shorter? It does not need to be so long. See [here](https://github.com/skypyproject/skypy/blob/df2536974c122dd72498f8d146913c14c832833b/skypy/galaxies/stellar_mass.py#L51).

philipp128 commented on skypy/galaxies/velocity_dispersion.py:
```diff
@@ -0,0 +1,69 @@
+    The probability distribution function :math:`p(\sigma)` for redshift :math:`\sigma`
+    is given by Choi et al. [1]_ as
+
+    .. math::
```
```suggestion
    .. math::
    
```

philipp128 commented on skypy/galaxies/velocity_dispersion.py:
```diff
@@ -0,0 +1,69 @@
+import scipy.special as special
+
+
+def schecter_vdf(vd_min, vd_max, resolution=100, size=None, scale=1.):
```
I also do not know if there are models of the velocity dispersion that allows `alpha`, `beta` or `vd_star` to be redshift dependent. If this is the case please proceed like [here](https://github.com/skypyproject/skypy/blob/df2536974c122dd72498f8d146913c14c832833b/skypy/galaxies/stellar_mass.py#L15) and [here](https://github.com/skypyproject/skypy/blob/df2536974c122dd72498f8d146913c14c832833b/skypy/galaxies/luminosity.py#L16).

philipp128 commented on skypy/galaxies/tests/test_velocity_dispersion.py:
```diff
@@ -0,0 +1,25 @@
+    size = 100
+    samples = velocity_dispersion.schecter_vdf(vd_min, vd_max, size=size)
+
+    # test output is within limits and sizee
```
```suggestion
    # test output is within limits and size
```

philipp128 commented on skypy/galaxies/tests/test_velocity_dispersion.py:
```diff
@@ -0,0 +1,25 @@
+    assert np.all(samples >= vd_min)
+    assert np.sum((vd_min <= samples) & (samples <= vd_max)) == size
+
+    # test output shape when scale is scalar
```
Scale should not be an input parameter. Replace this test for the shape with shape tests as done [here](https://github.com/skypyproject/skypy/blob/df2536974c122dd72498f8d146913c14c832833b/skypy/galaxies/tests/test_stellar_mass.py#L23) and [here](https://github.com/skypyproject/skypy/blob/df2536974c122dd72498f8d146913c14c832833b/skypy/galaxies/tests/test_luminosity.py#L7).

philipp128 commented on skypy/galaxies/tests/test_velocity_dispersion.py:
```diff
@@ -0,0 +1,25 @@
+    samples = velocity_dispersion.schecter_vdf(vd_min, vd_max, scale=scale)
+    assert np.shape(samples) == np.shape(scale)
+
+    # Test output shape when scale is an array
```
See above

philipp128 commented on skypy/galaxies/tests/test_velocity_dispersion.py:
```diff
@@ -0,0 +1,25 @@
+    # test output is within limits and sizee
+    assert np.all(samples <= vd_max)
+    assert np.all(samples >= vd_min)
+    assert np.sum((vd_min <= samples) & (samples <= vd_max)) == size
```
This test is unnecessary if you test for the shape. See below

philipp128 commented on skypy/galaxies/tests/test_velocity_dispersion.py:
```diff
@@ -0,0 +1,25 @@
+from skypy.galaxies import velocity_dispersion
```
```suggestion
```

philipp128 commented on skypy/galaxies/tests/test_velocity_dispersion.py:
```diff
@@ -0,0 +1,25 @@
+import numpy as np
+
+
+def test_schecter_vdf():
```
```suggestion
def test_schecter_vdf():
    from skypy.galaxies.velocity_dispersion import schechter_vdf
```

philipp128 replied:
You should also add a test of the sampling itself. See main comment.

Vikramb1 commented on skypy/galaxies/velocity_dispersion.py:
```diff
@@ -0,0 +1,69 @@
+    def vdf_func(x):
+        return 8e-3*(x/161)**2.32*np.exp(-(x/161)**2.67)*(2.67/special.gamma(2.32/2.67))*(1/x)
+
+    pdf = vdf_func(lnx)
+    cdf = pdf  # in place
+    np.cumsum((pdf[1:]+pdf[:-1])/2*np.diff(lnx), out=cdf[1:])
+    cdf[0] = 0
+    cdf /= cdf[-1]
+
+    t_lower = np.interp(vd_min, lnx, cdf)
+    t_upper = np.interp(vd_max, lnx, cdf)
+
+    u = np.random.uniform(t_lower, t_upper, size=size)
+    lnx_sample = np.interp(u, cdf, lnx)
+
+    return lnx_sample * scale
```
Hi, Philipp. I am not entirely sure how to do this. The velocity dispersion Schechter function is a modified version of the normal Schechter function, so it has a part which is `\exp[-{(\frac{\sigma}{{\sigma}_{*}})}^{\beta }]`, which means there is a beta parameter raised inside the exponent, which I think skypy's schechter function does not support. Perhaps I am misunderstanding.

 Also, in terms of adding an option for setting alpha and beta, do I leave the default as the values calculated in Choi et al.? E.g. `def schechter_vdf(alpha=2.32, ....`.

philipp128 replied:
To the first:
This is not a problem. You can re-parametrise and bring it into the same form since beta is a constant. If you start from equation 4 from the paper,
$$\phi(\sigma)\mathrm{d}\sigma = \phi_\* (\sigma/\sigma_\*)^\alpha \cdot e^2$$

philipp128 replied:
To the second:
No don't include the defaults of the Choi paper.

philipp128 replied:
Ah you are right with the exponent. This was my mistake. Sorry and thanks.
Your approach was already correct. So if we start with 
$$\phi(\sigma) \mathrm{d}\sigma = \phi_\* \left(\frac{\sigma}{\sigma_\*}\right)^\alpha \exp\left[-\left( \frac{\sigma}{\sigma_\*} \right)^\beta\right] \frac{\beta}{\Sigma(\alpha/\beta)} \frac{1}{\sigma} \mathrm{d}\sigma,$$
we can define
$$x = \left( \frac{\sigma}{\sigma_\*} \right)^\beta.$$
Taking the derivative we get
$$\frac{\mathrm{d}x}{\mathrm{d}\sigma} = \beta x \frac{1}{\sigma}$$
and
$$\mathrm{d}\sigma = \frac{\sigma}{\beta x}\mathrm{d}x.$$
Plugging this into the first equation we have
$$\phi(x)\mathrm{d}x = \phi_\* \left(\frac{\sigma}{\sigma_\*}\right)^\alpha \exp[-x] \frac{\beta}{\Sigma(\alpha/\beta)} \frac{1}{\sigma} \frac{\sigma}{\beta x} \mathrm{d}x = \frac{\phi_\*}{\Sigma(\alpha/\beta)} \frac{1}{x} \left(x^{\frac{1}{\beta}}\right)^\alpha \exp[-x] \mathrm{d}x = \frac{\phi_\*}{\Sigma(\alpha/\beta)} x^{\frac{\alpha}{\beta} - 1} \exp[-x] \mathrm{d}x.$$
Now we can define $\alpha'=\alpha/\beta - 1$ and get
$$\phi(x) = \frac{\phi_\*}{\Sigma(\alpha/\beta)} x^{\alpha'} \exp[-x].$$

I hope I didn't do any other mistake. Please correct me if this is the case.

Vikramb1 replied:
Thank you very much. I have implemented what I think you mean, but I am not sure if it is right. To test it against my own sampling, how do I normalise my function? Also, I have omitted the constant

$$\frac{\phi_*}{\Sigma(\alpha/\beta)}$$

Is this the correct thing?

philipp128 replied:
Thanks. Omitting the normalisation is fine in this case because in `random.schechter()` we calculate the cdf as a cumulative sum of the pdf and normalise it afterwards. 

To test if everything is correct we have the unit tests. You can do an independent sampling for a set of parameters and compare it to the sampling with this implementation. For this, you can use the implementation that you had initially. That would be a very good test. However, I took a first quick look at the new commits. Pay attention to the boundaries `vd_min` and `vd_max`. You need to change them to an `x_min` and `x_max`.

If you could add the unit test (you can find an example [here](https://github.com/Vikramb1/skypy/blob/df2536974c122dd72498f8d146913c14c832833b/skypy/galaxies/tests/test_stellar_mass.py#L53)) I can take another look early next week.

Vikramb1 replied:
Sure. I have a quick question about testing sampling. I have created a package some time ago: https://github.com/Vikramb1/Lens_population_sampler which I can import and test sampling from, but I would assume you would prefer the implementation without the importing of other packages, right?

philipp128 replied:
Yes, without other packages please

philipp128 commented on skypy/galaxies/velocity_dispersion.py:
```diff
@@ -0,0 +1,66 @@
+
+    .. math::
+
+        {dn}={\phi }_{*}{(\frac{\sigma}{{\sigma}_{*}})}^{\alpha}
+        {exp}[-{(\frac{\sigma}{{\sigma}_{*}})}^{\beta}]
+        \frac{\beta}{{\Gamma}(\alpha/\beta)}\frac{1}{\sigma},
```
```suggestion
        \mathrm{d}\phi(\sigma) = \phi _{*} \left(\frac{\sigma}{\sigma_{*}}\right)^{\alpha}
        \exp\left[- \left(\frac{\sigma}{\sigma_{*}}\right)^{\beta}\right]
        \frac{\beta}{\Sigma(\alpha/\beta)}\frac{1}{\sigma} ,
        
     where :math:`\Sigma` is ... , :math:`\sigma_*` is ... , :math:`\phi_*` is ... and :math:`\alpha` and :math:`\beta` are free parameters.
```
Fill the gaps please.

philipp128 replied:
Thanks for the added explanation. Could you maybe also clean up the LaTeX code a bit? I.e. removing unnecessary curly brackets etc and please keep consistent with the equation from the paper. Instead `dn` write `phi = ...` and use `Sigma` instead of `Gamma`. See my suggestions.

Vikramb1 replied:
I am a bit confused by what Sigma is. In the paper there is no mention of it. However in this paper: https://iopscience.iop.org/article/10.1088/0004-637X/811/1/20/pdf, equation 3 uses the modified schechter function and  uses the Gamma function. What should I add as to what Sigma is?

philipp128 replied:
Hmm ok. In this case I guess it is the gamma function. I think it better to use `Gamma` but still mention that it is the gamma function.

Vikramb1 replied:
Got it.

philipp128 commented on skypy/galaxies/velocity_dispersion.py:
```diff
@@ -0,0 +1,57 @@
+
+        \phi = \phi_* \left(\frac{\sigma}{\sigma_*}\right)^\alpha
+        \exp\left[-\left( \frac{\sigma}{\sigma_*} \right)^\beta\right] 
+        \frac{\beta}{\Gamma(\alpha/\beta)} \frac{1}{\sigma} \mathrm{d}\sigma
```
Please check the rendering of the equation and the mathematical expressions in the docs. It does not generate correctly.

philipp128 commented on skypy/galaxies/velocity_dispersion.py:
```diff
@@ -0,0 +1,57 @@
+        \exp\left[-\left( \frac{\sigma}{\sigma_*} \right)^\beta\right] 
+        \frac{\beta}{\Gamma(\alpha/\beta)} \frac{1}{\sigma} \mathrm{d}\sigma
+
+        where :math:`\Gamma` is the gamma function, :math:`\sigma_*` is the charactersitic velocity dispersion, :math:`\phi_*` 
```
```suggestion
      where :math:`\Gamma` is the gamma function, :math:`\sigma_*` is the characteristic velocity dispersion, :math:`\phi_*` 
```
I think the wrong rendering comes from the wrong indentation. Please verify that this aligned with the other text.

philipp128 commented on skypy/galaxies/velocity_dispersion.py:
```diff
@@ -0,0 +1,57 @@
+def schechter_vdf(alpha, beta, vd_star, vd_min, vd_max, size=None, resolution=1000):
+    r"""Sample velocity dispersion of elliptical galaxies in the local universe following a Schecter function.
+
+    Parameters
```
Please add a description for all input parameters.

philipp128 reviewed (changes requested):
Thanks @Vikramb1, it looks really good now but I still have some minor things I'd ask you to check again. See the corresponding comments.

Also, can you check for code style using flake8 before the next commit. The corresponding test is failing at the moment. Also, please check if the documentation is rendering nicely. You can do both locally. Instructions are in the contributing guidelines.

philipp128 reviewed (commented):
Looks good to me now. @rrjbca or @itrharrison could one of you take a second look?

itrharrison reviewed (approved):
This looks very good to me, thank you very much for your contributions both!
